### ISSUE TITLE
cast to non ptr UncheckedArray does not produce an error or warning

### ISSUE THREAD

**Issue by cooldome** [nim-lang/Nim#9403]
Used to compile:
```
type
    MyObj = ref object
      len: int
      val: UncheckedArray[float]
  
   
proc spot(x: MyObj): int64 =
  result = cast[UncheckedArray[int64]](x.val)[0]
  
   
proc newMyObj(len: int): MyObj =  
  unsafeNew(result, sizeof(result[]) + len * sizeof(float))
  result.len = len
    
echo spot(newMyObj(2))
```

Now generated C fails to compile with error:
```
nimcache\regression1_d\compiler_regression1.c: In function 'spot_a7qWq7kPv9bXGgmHKgB6vbw':
nimcache\regression1_d\compiler_regression1.c:140:12: error: cast specifies array type
  result = ((tyUncheckedArray_jpBG9aCOjd7lf9cMZpCfZm9bw) ((*x).val))[((NI) 0)];
```
simple workarounds exists

**Comment by cooldome** [comment]
I don't mind a solution that will disallow such kind of casts in semantic pass, they are of minor usefulness, but still can be useful.

I do not agree with the statement:
  A bare UncheckedArray (not prefixed with ref or 
  ptr) should never be used IMO.

Having unchecked array as tail in your object is petty common idiom in C for example. This is also how current GC seq is implement in Nim

**Comment by LemonBoy** [comment]
> I do not agree with the statement:
>> A bare UncheckedArray (not prefixed with ref or ptr) should never be used IMO.

I meant outside of an object declaration, I'm well aware of the extensible-array trick and am also aware that the compiler should be more strict about those (see #9275)

> I don't mind a solution that will disallow such kind of casts in semantic pass, they are of minor usefulness, but still can be slightly useful.

Not really, if you want to cast an array/seq/whatever to a `UncheckedArray` you go trough its address.
If you write something like this you're getting a full-blown copy of the `q` array in `z`.

```nim
var q = [1,2,3]
var z = cast[array[3,int]](q)
echo z
```

If you try to apply the same logic to a `UncheckedArray` you can easily see that we don't know how many elements we have and we'd end up copying a single element instead (but that's an implementation detail).
The bottom line is, I'm gonna add a check in `isCastable` in order to make this an error.
